### DISCUSSION TITLE
I have question for memory usage about querynode!

### DISCUSSION THREAD

**Discussion by hjlee9182** [milvus-io/milvus#31156] [Q&A and General discussion] [answered]
milvus version: 2.3.10
querynode : 1
collection info: 1M, 768dim

1. I did load and release same collection with ivf_flat index. Every load and release memory is different. Can you explain this?
2.01GB -> (load) -> 5.02GB -> (release) ->1.88GB -> (load) -> 4.93GB -> (release) -> 1.83GB -> (load) -> 4.96GB -> 1.77GB
![image](https://github.com/milvus-io/milvus/assets/46337218/cbfedc48-41bc-4246-b9ac-105173afe3a0)

2. I measured memory usage on load for each index. Is there a memory usage calculation formula for each index?
```
IVF_FLAT: 3.04GB
IVF_SQ8: 0.81GB
FLAT: 2.78GB
DISKANN: 0.64GB
```

3. Why diskann need memory when load collection?
4. when I search to collection with diskann index. memory is increasing. I guess that's cache. Is there cache memory limit?
5. I chcked collection with diskann index was loaded, disk usage is increase. How to calculate?
before
<img width="370" alt="image" src="https://github.com/milvus-io/milvus/assets/46337218/28a7d75b-f96f-425f-8183-ab99264a1ee1">

after
<img width="393" alt="image" src="https://github.com/milvus-io/milvus/assets/46337218/e3c19277-0164-41c9-be3e-e08bbe72d41d">

**Comment by xiaofan-luan** [comment, marked as answer]
> milvus version: 2.3.10 querynode : 1 collection info: 1M, 768dim
> 
> 1. I did load and release same collection with ivf_flat index. Every load and release memory is different. Can you explain this?
>    2.01GB -> (load) -> 5.02GB -> (release) ->1.88GB -> (load) -> 4.93GB -> (release) -> 1.83GB -> (load) -> 4.96GB -> 1.77GB
>    ![image](https://private-user-images.githubusercontent.com/46337218/311448159-cbfedc48-41bc-4246-b9ac-105173afe3a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAwMDY4MzAsIm5iZiI6MTcxMDAwNjUzMCwicGF0aCI6Ii80NjMzNzIxOC8zMTE0NDgxNTktY2JmZWRjNDgtNDFiYy00MjQ2LWI5YWMtMTA1MTczYWZlM2EwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA5VDE3NDg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3YWVmMWUwOTQ0ZjE4MWZmMGRmY2Y1ODBjZjdkZWVkMmUzMDg1ODM1MWExNWIyOWZhM2U2NjhlMDkzMjRhZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1mVIJe9SeooUQB1i0IspDB2Y3Opo8ORX0tpND3cylhM)
> 2. I measured memory usage on load for each index. Is there a memory usage calculation formula for each index?
> 
> ```
> IVF_FLAT: 3.04GB
> IVF_SQ8: 0.81GB
> FLAT: 2.78GB
> DISKANN: 0.64GB
> ```
> 
> 3. Why diskann need memory when load collection?
> 4. when I search to collection with diskann index. memory is increasing. I guess that's cache. Is there cache memory limit?
> 5. I chcked collection with diskann index was loaded, disk usage is increase. How to calculate?
>    before
> 
> <img alt="image" width="370" src="https://private-user-images.githubusercontent.com/46337218/311449760-28a7d75b-f96f-425f-8183-ab99264a1ee1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAwMDY4MzAsIm5iZiI6MTcxMDAwNjUzMCwicGF0aCI6Ii80NjMzNzIxOC8zMTE0NDk3NjAtMjhhN2Q3NWItZjk2Zi00MjVmLTgxODMtYWI5OTI2NGExZWUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA5VDE3NDg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4YWZiMmUzNDYxNjhlOWNjY2Y1YmEyMjg1ZDYxMGRlMTdmY2I2NmEzNDNmOTkyN2NjZDBiMTM3MWUxMjkxZWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tM-ue1jzozeYDMHVe1veXkLg47qV0O_f7TUDbOkPj44">
> after <img alt="image" width="393" src="https://private-user-images.githubusercontent.com/46337218/311449767-e3c19277-0164-41c9-be3e-e08bbe72d41d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAwMDY4MzAsIm5iZiI6MTcxMDAwNjUzMCwicGF0aCI6Ii80NjMzNzIxOC8zMTE0NDk3NjctZTNjMTkyNzctMDE2NC00MWM5LWJlM2UtZTA4YmJlNzJkNDFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA5VDE3NDg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxMWE2NjAyYzA1N2NmMGIwNmUxYzMxOGVjYmQ3ODBkMmFjNTY0OWI2YmIzZTVmMTFjZDJlMGM1NzhlNTg4ODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OTOlw_-iuoB6ecSWo_3rjrUmYLbuy-aAQsFwkuduWdU">

Hi @hjlee9182 
you did very detailed test! 
For the memory fluctuation, I guess that is due to go GC and Jemlloc reserved some memory. From your test I saw that memory is not increasing so I guess generally there are no leakage. Tune the jemalloc memory purge parameter could probably help on your verification

2. SQ8 is roughly 1/4th of IVFFlat, FLAT is slightly less than FLAT, diskANN is roughly 1/6th but it kind like depend on your parameters. you can do 1/12th or even less if you tune the parameters.

4. Diskann has a graph based index in memory, but it is just quantized so it saves memory. That is saying you still need memory to host your disk. 

5. DiskANN should has a cache. but it should be configurable to disable to tune it to smaller

**Reply by hjlee9182** [reply]
@xiaofan-luan 
Thank you for answering!

> DiskANN should has a cache. but it should be configurable to disable to tune it to smaller

You mean SearchCacheBudgetGBRatio config in docs?
https://milvus.io/docs/disk_index.md#DiskANN-related-Milvus-configurations

> 5. I chcked collection with diskann index was loaded, disk usage is increase. How to calculate?

Can you answer question 5 in the text? :)

**Comment by yhmo** [comment]
- DiskANN is "in-memory PQ index" +  "on-disk Vamana graph", read this article: https://milvus.io/blog/2021-09-24-diskann.md

- DiskANN index is configured in the milvus.yaml: https://milvus.io/docs/disk_index.md#DiskANN-related-Milvus-configurations
MaxDegree,  SearchListSize, PQCodeBugetGBRatio, SearchCacheBudgetGBRatio affect its on-disk size.
PQCodeBugetGBRatio and SearchCacheBudgetGBRatio affect its in-memory size.

- When you load a collection with diskann, it loads the in-memory index into memory. When you search the collection, it first finds the candidates from in-memory index and then reads the real data of candidates from disk. It makes sense the memory usage is not a const value. The memory usage is not always increase without limited. No need worry.

**Comment by xiaofan-luan** [comment]
5. the increate of your disk usage might be caused by
1. your wal if you are deploying a standalone and use rocksmq
2. compaction triggered and new file deleted 

All of the two will be expired after a few days

**Comment by hjlee9182** [comment]
IS IVF_SQ8 also using disk?

before
<img width="470" alt="image" src="https://github.com/milvus-io/milvus/assets/46337218/c9fe7bb6-210b-4119-a0e0-f17b87ab896f">


after
<img width="392" alt="image" src="https://github.com/milvus-io/milvus/assets/46337218/7e0c1729-3711-403d-9e07-1ab33cdc7cd7">

**Reply by yhmo** [reply]
All indexes are persisted in disk.
When you call collection.load(), the index is loaded into the memory of query nodes.

**Reply by hjlee9182** [reply]
But when IVF_FLAT, FLAT Index loaded, disk usage is not increase.

**Reply by yhmo** [reply]
Milvus manages data by segments. The size of each segment is expected to be 512MB(by default).

When you insert data, data is first stored into Pulsar/Kafka as write-ahead-log, you will see disk usage increased.

Data node and query node subscribe the data from Pulsar/Kafka/RocksDB, the data is accumulated in a "growing segment".

Once the size of growing segment exceeds a threshold(110MB by default) or you manually call collection.flush(), the growing segment is converted to a sealed segment and persists to storage. you will see disk usage increased.

Once a sealed segment is persisted, the index node is notified and starts to build index for this segment. Usually, a 100Mb segment costs half or 1 minute to build index. you will see disk usage increased.

At the same time, if there are several sealed segments generated, the data node will trigger a "compaction" task to merge them to be a large sealed segment. you will see disk usage increased.

A growing segment can be searched with brute-force.
A sealed segment without index can be searched with brute-force.
A sealed segment with index is searched with index.

So, even you can search the collection, there might be index tasks and compactions running in the background.

For example:
1. you insert 100MB data from 11:25 to 11:30, the data is stored in Pulsar/Kafka/RocksDB, disk usage is 100MB.
2. a sealed segment_A(100MB) is generated at 11:30, the disk usage is 100+100 = 200MB
3. the index node begins to build index for segment_A, after 1 minute, at 11:31 the disk usage is 200MB + 100MB(according to index type, FLAT/IVF_FLAT/HNSW 100MB, IVF_SQ8/IVF_PQ 30MB) = 300MB
4. you insert 100MB data from 11:30 to 11:35, the data is stored in Pulsar/Kafka/RocksDB, disk usage is 300+100=400MB.
5. a new sealed segment_B(100MB) is generated at 11:35, the disk usage is 400 +100 = 500MB
6. the index node begins to build index for segment_B, the task is finished at 11:36, the disk usage is 500+100 = 600MB
7. at the same time, the data node triggers a compaction to merge segment_A and segment_B to a new segment_C, the size of segment_C is 100+100=200MB. the compaction task is finished at 11:37, the disk usage is 600 + 200 = 800MB
8. the index node begins to build index for segment_C, the task is finished at 11:39, the disk usage is 800 + 200 = 1GB
9. the segment_A and segment_B and their indexes are marked as "deleted", and wait GC process to permanently clean frpm disk. by default, GC process is triggerd per 1~2 hour. After GC, the disk usage is 1GB - 100 -100 -100 -100 = 600MB.
10. the data in Pulsar/Kafka will be kept in storage for 1 ~ 3 days, after 1 ~ 3 days, the write-ahead-log data is cleaned, the disk usage is 600 - 200 = 400MB.

So, you only insert 200MB totally, but the disk usage is increase to 1GB in 10+ minutes.
Milvus is a complicated distributed system, there are lots of async tasks running in the background, to process data and requests. All behaviors have their reasons and logic.
